src/libmeasurement_kit/common/utils_impl.hpp

```diff
@@ -0,0 +1,63 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+#ifndef SRC_LIBMEASUREMENT_KIT_COMMON_UTILS_IMPL_HPP
+#define SRC_LIBMEASUREMENT_KIT_COMMON_UTILS_IMPL_HPP
+
+#include "../common/utils.hpp"
+#include <cstdio>
+#include <measurement_kit/common.hpp>
+
+namespace mk {
+
+/*
+ * There are many ways in C++ to slurp the whole content of a file into a
+ * C++ string. The following is my favourite. Compared to other methods,
+ * not only this seems fastest, but also it provides greater control over
+ * errors that can occur. Thus mocking is simpler.
+ *
+ * See <http://stackoverflow.com/questions/116038/what-is-the-best-way-to-read-an-entire-file-into-a-stdstring-in-c>.
+ */
+template <typename T, MK_MOCK_NAMESPACE(std, fopen),
+          MK_MOCK_NAMESPACE_SUFFIX(std, fseek, 1),
+          MK_MOCK_NAMESPACE(std, ftell),
+          MK_MOCK_NAMESPACE_SUFFIX(std, fseek, 2),
+          MK_MOCK_NAMESPACE(std, fread), MK_MOCK_NAMESPACE(std, fclose)>
+ErrorOr<std::vector<T>> slurpv_impl(std::string p) {
+    FILE *filep = std_fopen(p.c_str(), "rb");
+    if (filep == nullptr) {
+        return FileIoError();
+    }
+    if (std_fseek_1(filep, 0, SEEK_END) != 0) {
+        std_fclose(filep);
+        return FileIoError();
+    }
+    // Note: ftello() might be better for reading very large files but
```

Comment:
I am not sure. In theory, I'd rather have the system tell me I'm using too much memory by having `malloc()` return `NULL`. Fixed constraints in general give a false sense of security and possibly limit resources unreasonably. Consider that by adding fixing constraints we're basically trying to do the job of the system with much less information. In general, that's a no-no.

But, on the other hand, we currently use `new` in many places and `new` can `throw` and we don't care about `std::bad_alloc()`. Furthermore, AFAIK under Linux `malloc()` never returns `NULL` because `Linux` has an optimistic memory allocation policy. That's basically my argument above except that it's applied to the `malloc()` libc function, which is less well versed that the kernel memory manager to judge and guess whether we're short of memory pages. The disadvantage of Linux policy however is that you basically `segfault` when you ask for memory and the system does not have a memory page available for you. That is, you cannot handle it gracefully. To be honest, I really don't like this Linux behavior and I prefer the behavior of BSD systems. (Note: what I am saying here is what I recall from some years ago, perhaps this changed.)

Bottom line is: we don't know in advance whether the system is going to tell us that we're running out of memory and in many cases we don't handle `bad_alloc()` for this reason. I still do not see a reason to add a fixed limit to the function, tho. That would still sound like arbitrary. I'd rather add a parameter allowing the caller to specify the maximum acceptable size of the file according to his judgement. This way it would not sound arbitrary to the user of the function.

Does it make sense?